src/server/src/main/java/io/cassandrareaper/storage/cassandra/SchemaMigrationLock.java

```diff
@@ -0,0 +1,172 @@
+import com.datastax.driver.core.utils.UUIDs;
+import jersey.repackaged.com.google.common.base.Preconditions;
+import org.cognitor.cassandra.migration.Database;
+import org.cognitor.cassandra.migration.MigrationRepository;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class SchemaMigrationLock {
+
+  private static final UUID SCHEMA_MIGRATION_LOCK_ID = UUIDs.startOf(1);
+  private static final Logger LOG = LoggerFactory.getLogger(SchemaMigrationLock.class);
+  private static PreparedStatement lockSchemaMigrationPrepStmt;
+  private static PreparedStatement unlockSchemaMigrationPrepStmt;
+  private static PreparedStatement renewLockSchemaMigrationPrepStmt;
+  private final VersionNumber version;
+  private final Session session;
+  private final ReaperApplicationConfiguration config;
+
+  public SchemaMigrationLock(VersionNumber version, Session session, ReaperApplicationConfiguration config) {
+    this.version = version;
+    this.session = session;
+    this.config = config;
+    if (0 >= VersionNumber.parse("2.2").compareTo(version)) {
+      // no system_distributed keyspace before 2.2
+      lockSchemaMigrationPrepStmt = session.prepare("insert into system_distributed.parent_repair_history"
+          + " (parent_id, started_at)"
+          + " values(?, dateOf(now())) IF NOT EXISTS USING TTL ?");
+      unlockSchemaMigrationPrepStmt = session.prepare("DELETE FROM system_distributed.parent_repair_history"
+          + " WHERE parent_id = ?");
+      renewLockSchemaMigrationPrepStmt = session.prepare("update system_distributed.parent_repair_history USING TTL ?"
+          + " set started_at = dateOf(now()) where parent_id = ?");
+    }
+  }
+
+  public synchronized boolean isSchemaMigrationRequired(com.datastax.driver.core.Cluster cassandra)
+      throws ReaperException {
+    Preconditions.checkArgument(!config.isInSidecarMode() || config.getEnableConcurrentMigrations(),
+        "Running in sidecar mode and concurrent migrations are not enabled.");
+
+    Optional<Boolean> migrationRequired = Optional.empty();
+    while (!migrationRequired.isPresent()) {
+      try (Database database = new Database(cassandra, config.getCassandraFactory().getKeyspace())) {
+        int currentVersion = database.getVersion();
+        MigrationRepository migrationRepo = new MigrationRepository("db/cassandra");
+        if (currentVersion < migrationRepo.getLatestVersion()) {
+          if (lockSchemaMigration()) {
+            // We own the lock and a migration is required
+            migrationRequired = Optional.of(true);
+          } else {
+            // Migration is required but another reaper instance owns the lock
+            LOG.info("Cannot grab the lock on schema migration. Waiting for it to be released...");
+            try {
+              Thread.sleep(10000);
+            } catch (InterruptedException e) {
+              // TODO Auto-generated catch block
+              throw new ReaperException(e);
```

Comment:
swallow or `IllegalStateException`

Reply:
see what i've done in https://github.com/thelastpickle/cassandra-reaper/commit/eb7a757ba38be468f84c191197be2f9ba915ebfe